show only versions newer than NVM_MIN if set

ryenus opened the pull request:
This is to fix #3250.

ryenus commented:
For example, running `NVM_MIN_VER=18 nvm ls-remote --lts` when v14 and v16 are installed:

<img width="360" alt="image" src="https://github.com/nvm-sh/nvm/assets/610161/720f0ca3-f7ab-4b55-9444-a7fc5df8a00c">

ljharb reviewed (commented):
I like that it's done in awk :-) it'd be great to benchmark the difference to make sure this isn't slowing anything down meaningfully.

this will need tests, and also it'd be good to add a `--min` option to all the things that talk to the mirror (`nvm ls-remote`, `nvm install`, `nvm version-remote`) and then maybe `nvm ls` for good measure (altho that can be a follow-on if you like)

ryenus commented:
Performance-wise there's not really much change or overhead, you can trust that awk is fast :-) Sometimes it might even run faster, esp. when many older versions are skipped. 

Meanwhile the other sibling functions and even the tests are something I'm not really familiar with, may you can chime in here? @ljharb

ryenus commented:
@ljharb just added a test, however Travis CI seems stuck, any insight?

ryenus commented:
BTW, I had to adjust the travis job a bit, due to the error

    pyenv: /opt/pyenv/versions/2.7.18 already exists

See the screenshot below (https://app.travis-ci.com/github/nvm-sh/nvm/jobs/623514514):

<img width="968" alt="image" src="https://github.com/nvm-sh/nvm/assets/610161/92ddc7ee-1560-44e6-b35b-242553aa3c56">

ljharb commented on nvm.sh:
```diff
@@ -1820,12 +1820,14 @@ nvm_print_versions() {
 
   command awk \
-    -v remote_versions="$(printf '%s' "${1-}" | tr '\n' '|')" \
+    -v remote_versions="$(printf '%s' "${1-}" | tr '\n' '|')" -v min_ver="${NVM_MIN:-v0}" \
```
is defaulting it to `v0` the right choice, as opposed to skipping the logic entirely?

ryenus replied:
Surely we could do that, meanwhile versions like `v0`, `v0.0`, and `0.0.0` (with the prefix `v`), are all legit, therefore, we have the `filter_on` flag which is deduced with this:

    filter_on = (vcmp("v0.0.0", min_ver) != 0);

And given that the versions list is already sorted, once we encounter the version greater or equal to the min version specified, we turn off the filtering logic and show all versions onward.

ljharb commented on nvm.sh:
```diff
@@ -1863,8 +1872,7 @@ BEGIN {
 
-    padding = (!has_colors && is_installed) ? "" : "  ";
-
+    padding = (is_installed && !has_colors) ? "" : "  ";
```
nbd, but why flip these conditions?

ryenus replied:
Not really flipping, just swapping the LHS and RHS of the `&&` operator. This slightly improves the clarity and makes it clear that only `!has_colors` is the negating part. This is something I should have done in the first place.

ljharb commented on test/fast/Unit tests/nvm_print_versions:
```diff
@@ -0,0 +1,139 @@
+[ "_$OUTPUT" = "_$EXPECTED_OUTPUT" ] || die "nvm_print_versions did not output all expected versions; got $OUTPUT"
+
+
+# versions lower than 18 should be filtered out, but v16.20.2 should be kept since it's installed
+OUTPUT="$(NVM_NO_COLORS=1 NVM_MIN=18 nvm_print_versions "$(nvm_remote_versions)" | sed -r 's/^[ \t]+//')"
```
what if you have v16.20.1 installed, for example, and you still want to see that v16.20.2 is both available, and the latest LTS from that line?

maybe anything that's installed _plus_ `^` of it should still be shown?

or maybe, we should skip the "keep if installed" logic entirely?

ryenus replied:
I think the chosen behavior is slightly more preferable, due to the fact that `nvm ls-remote` has always been highlighting the installed versions, so that people can intuitively compare the versions they installed, to the newer versions available. And that kind of intuition is needed for people to decide whether or when to install a new version, which is much nicer than having to run both `nvm ls-remote` and `nvm ls`.

ljharb replied:
ok - in that case, setting a MIN_VER of 18, with 16.20.1 installed as the only 16, should still show 16.20.2 as an installable version.

ryenus replied:
Hmmm, isn't that a bit too far? If we do that then it means the min version is effectively the oldest installed version, hence defeating the point of specifying a min version. Nevertheless I agree it's a bit tricky to decide on which is the sweet spot.

Another option could be that we first let it out and see the user perception from the community.

ljharb replied:
In my scenario, 17 still wouldn’t be displayed. So, it’d display all installed versions, anything that can be safely upgraded to, as well as the min version and above.

Unfortunately community perception and feedback is in practice almost never sufficient enough to give an accurate picture of the actual community reception - just the most vocal members - so it’s important the project is happy with the way it works initially.

ryenus replied:
Just clarify, for an installed version, with both newer patch versions and newer minor versions, do you expect to see just the patch version updates? or both?

ljharb replied:
I’d expect both - everything up to and not including the next major

ryenus replied:
Done, see https://github.com/nvm-sh/nvm/pull/3277/commits/477d22917

Now when we encounter a version that's installed, we make sure to not skip all following versions having the same major version.

ljharb reviewed (commented):
It occurs to me that

ryenus commented:
https://app.travis-ci.com/github/nvm-sh/nvm/jobs/624025129#L797

```
+rm -fR v* src alias test/test-xz
  ✗ Running 'nvm unload' should unset all function and variables
+BEFORE=./before.tmp
```

@ljharb any idea if the failing Travis test is related?

ryenus commented:
As we can see from the below screenshot, for v16.15.1, now all of its updates are listed:

<img src='https://github.com/user-attachments/assets/a4463aeb-e877-4041-8dcd-93b6d845a509' width=350/>

ljharb commented on nvm.sh:
```diff
@@ -3024,6 +3036,7 @@ nvm() {
         nvm_echo '  nvm ls-remote [<version>]                   List remote versions available for install, matching a given <version> if provided'
         nvm_echo '    --lts                                     When listing, only show LTS (long-term support) versions'
         nvm_echo '    --lts=<LTS name>                          When listing, only show versions for a specific LTS line'
+        nvm_echo '    --min=<version>                           When listing, only show versions greater than or equal to <version>, including minor/patch updates for installed versions'
```
should this same option work for `nvm ls` as well?

ryenus replied:
That would be an no-op since the installed versions, which is what reported by `nvm ls`, are always listed.

ljharb commented on nvm.sh:
```diff
@@ -4000,6 +4013,7 @@ nvm() {
     ;;
     "ls-remote" | "list-remote")
       local NVM_LTS
+      local NVM_MIN="${NVM_MIN-}"
```
```suggestion
      local NVM_MIN
      NVM_MIN="${NVM_MIN-}"
```
this is needed for ksh

ryenus replied:
Ah, good to know!

Nevertheless the idea is to inherit `$NVM_MIN` from env, I'm not sure if that would still work with a change like this, maybe we need to introduce another variable as a workaround?

ljharb reviewed (commented):
This looks great, I'll play with it locally a bit this week.

ljharb commented:
When I run `nvm ls` locally with this PR, I get `empty` as the first item in the list.

When I run `nvm ls-remote --min=4` locally with this PR, all versions are displayed.

ryenus commented:
> When I run `nvm ls` locally with this PR, I get `empty` as the first item in the list.

Interesting, any idea where could this `empty` come from? Asking because here the change to `nvm_print_versions` only determines whether to skip certain versions or not, meanwhile all the versions are supplied by the caller.

May I ask what's the output of `nvm ls` when run from the master branch?

> When I run `nvm ls-remote --min=4` locally with this PR, all versions are displayed.

As a comparison, here's what I got with the same command:

<img width="294" alt="image" src="https://github.com/user-attachments/assets/4d23927a-3b00-41ee-bffe-049d76070da5">

Is it possible that the shell was still running the code anther branch, rather than that of the PR? Maybe check the output of `type nvm_print_versions`?

ljharb commented:
I'm certain that I'm using the code from this PR, yes.

I'm running this on a Mac; if you're not, then perhaps there's some platform-specific difference here?

ljharb commented:
hm, ok i did just figure out that i had a `$NVM_DIR/versions/node/empty` dir for some reason, so at least that's a red herring. Let me test again.

ljharb commented:
ok, testing again, and `nvm ls-remote --min=4` (or 20, or whatever) still has no effect for me on my Mac.

What OS are you using exactly?

ryenus commented:
> ok, testing again, and `nvm ls-remote --min=4` (or 20, or whatever) still has no effect for me on my Mac.
> 
> What OS are you using exactly?

Actually I'm also on mac :-) How about `NVM_MIN=4 nvm ls-remote`? Can you also try to enable `set -x` within `nvm_print_versions`?

ljharb commented:
ohhh wait, lol i think this is the actual feature i asked for - i have tons of node versions installed, going back a very long way:
```sh
$ NVM_MIN=4 nvm ls-remote | wc -l
     711
$ nvm ls-remote --min=4 | wc -l
     711
$ nvm ls-remote | wc -l
     790
```
in other words, it's filtering some things out, but with the additional logic i requested, it's still showing _me_ a lot of things i didn't expect. 🤔

ryenus commented:
> ohhh wait, lol i think this is the actual feature i asked for - i have tons of node versions installed, going back a very long way:
> 
> ```shell
> $ NVM_MIN=4 nvm ls-remote | wc -l
>      711
> $ nvm ls-remote --min=4 | wc -l
>      711
> $ nvm ls-remote | wc -l
>      790
> ```
> 
> in other words, it's filtering some things out, but with the additional logic i requested, it's still showing _me_ a lot of things i didn't expect. 🤔

Because you have many versions installed? Or Would you change mind? :P

BTW, typically I just run something like `NVM_MIN=18 nvm ls-remote --lts`.

ryenus commented:
@ljharb FYI, just rebased with some minor fix in tests.

ryenus commented:
@ljharb how is it going? anything to note?

ljharb commented:
No progress yet; I've been traveling. I'm prioritizing a bug fix on nvm before this PR, but it'll be next in line after that.

ljharb commented:
ok, so i just tried it out again.

if i do `nvm ls-remote --min=99`, i get all the versions hidden below the lowest one i have installed - v0.4.12 - but i get every version higher. So, for me, it's not really that useful. I assume that on an empty system, or one where all installed node versions are above the min, it actually works as intended.

I'm starting to wonder if perhaps my original concern - about versions being hidden - is better addressed by prepending a message like "Minimum version X set: N versions hidden", and going back to the simpler logic? (i'm super sorry if this means you're doing work to undo the work i already asked you to do :-/ unfortunately sometimes it takes time and first-hand experience to really get a sense of the DX/UX of a thing)

ryenus commented:
@ljharb, although in some cases it's somehow possible to have a broad range of node versions installed, but I'm sure not everyone would need to do that. I believe most people only need _**(a)**_ several (if not a few), and _**(b)**_ most recent, _**(c)**_ typically LTS versions, installed for the project(s) they work on. Like what's shown in [a previous comment][1]: 

> <img src='https://github.com/user-attachments/assets/a4463aeb-e877-4041-8dcd-93b6d845a509' width=320/>

And this is sort of the next step for the previous pr https://github.com/nvm-sh/nvm/pull/2827, that one has improved the performance of `nvm ls-remote`, by reducing the time needed for a complete `nvm ls-remote` from about 20s down to about 2s. This pr aims to also improve the ergonomics, by showing just what's really needed, like `798` (_before_) vs `38` (_after_) as shown below, nevertheless the time taken is also reduced as a side effect:

<table><tr><th>Unfiltered</th><th>Filtered</th></tr>
<tr><td>

```
# time nvm ls-remote --min=v0 | wc -l
798

real	0m1.253s
user	0m0.246s
sys	0m0.322s
```

</td>
<td>

```
# time nvm ls-remote --min=18 --lts | wc -l
38

real	0m0.545s
user	0m0.154s
sys	0m0.188s
```

</td></tr></table>

Meanwhile no worries, your initial proposal actually inspired me to think more about why people would need `nvm ls-remote`, or even nvm: it's because people would want to find newer versions of node delivering new features, patches or whatever improvements that are not available in currently installed versions. In that sense showing new patch versions definitely helps. That works like the program *automatically* detects a suitable min version based on installed ones, saving the need to figuring out the *closest* min version to specify, which is actually nice.

[1]: https://github.com/nvm-sh/nvm/pull/3277#issuecomment-2227304013

ryenus commented:
BTW, the error with `nvm --version` in the failed tests seems unrelated:
* https://github.com/nvm-sh/nvm/actions/runs/11406338964/job/31739908143?pr=3277
* https://github.com/nvm-sh/nvm/actions/runs/11406338964/job/31739908480?pr=3277

vincent89291 reviewed (commented):
Explicit

ryenus commented:
Hi, @ljharb, I've updated the code per your most recent feedback and also rebased.

So now `nvm ls-remote --min=<ver>` strictly lists versions newer than the min version specified. Hopefully we can move forward on this, thank you!

ryenus committed:
ls-remote: show only versions newer than NVM_MIN

add CLI option --min=<version>, fallback to env var "NVM_MIN" if set.
display a hint when version filtering is in effect

Other minor changes:
- fix: disable color in nvm_print_versions tests
- trim leading space in mock output due to eslint errors
- prefixed versions like v18 also work
- remove function ref duplication
- avoid inline initialization for ksh compatibility

Co-authored-by: Jordan Harband <ljharb@gmail.com>
